File: frontend/src/api/report.ts
import { api } from "./client";

export type ReportWeakNodeOut = {
  knowledge_node_id: string | null;
  knowledge_node_name?: string | null;
  wrong_count: number;
  total_count: number;
};

export type ReportOverviewOut = {
  subject_code: string | null;
  wrong_source_counts: Record<string, number>;
  weak_nodes: ReportWeakNodeOut[];
  self_test_trend: Array<{
    submission_id: string;
    paper_id: string;
    subject_code: string;
    total_score: number;
    created_at: string;
  }>;
};

export function fetchStudentReportOverview(params?: {
  subject_code?: string;
  trend_limit?: number;
}) {
  const qs = new URLSearchParams();
  if (params?.subject_code) qs.set("subject_code", params.subject_code);
  if (typeof params?.trend_limit === "number") qs.set("trend_limit", String(params.trend_limit));
  const suffix = qs.toString() ? `?${qs.toString()}` : "";
  return api<ReportOverviewOut>(`/student/report/overview${suffix}`);
}

File: backend/app/schemas/report.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class ReportWeakNodeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    knowledge_node_id: uuid.UUID | None
    knowledge_node_name: str | None = None
    wrong_count: int = Field(ge=0)
    total_count: int = Field(ge=0)


class ReportTrendPointOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    submission_id: uuid.UUID
    paper_id: uuid.UUID
    subject_code: str
    total_score: int
    created_at: datetime


class ReportOverviewOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    subject_code: str | None
    wrong_source_counts: dict[str, int]
    weak_nodes: list[ReportWeakNodeOut]
    self_test_trend: list[ReportTrendPointOut]
    recommendations: list[dict] = Field(default_factory=list)

